samza-core/src/main/java/org/apache/samza/operators/StreamGraphSpec.java

```diff
@@ -230,22 +249,44 @@ public String getNextOpId(OpCode opCode) {
    * @param <M> the type of messages in the intermediate {@link MessageStream}
    * @return  the intermediate {@link MessageStreamImpl}
    */
-  @VisibleForTesting
-  <M> IntermediateMessageStreamImpl<M> getIntermediateStream(String streamId, Serde<M> serde, boolean isBroadcast) {
+  public <M> IntermediateMessageStreamImpl<M> getIntermediateStream(String streamId, Serde<M> serde, boolean isBroadcast) {
     Preconditions.checkState(!inputOperators.containsKey(streamId) && !outputStreams.containsKey(streamId),
         "getIntermediateStream must not be called multiple times with the same streamId: " + streamId);
 
-    if (serde == null) {
-      LOGGER.info("Using default serde for intermediate stream: " + streamId);
-      serde = (Serde<M>) defaultSerde;
+    Serde streamSerde;
+    if (serde != null) {
+      streamSerde = serde;
+    } else {
+      if (defaultSystemDescriptor != null) {
+        LOGGER.info("Using default system's serde for intermediate stream: " + streamId);
+        streamSerde = (Serde) defaultSystemDescriptor.getSystemSerde().orElse(null);
+      } else {
+        LOGGER.warn("No serde defined for intermediate stream: " + streamId +
+            ". Key and message serdes specified for the job.default.system will be used.");
+        streamSerde = null;
+      }
     }
 
     if (isBroadcast) broadcastStreams.add(streamId);
-    boolean isKeyed = serde instanceof KVSerde;
-    KV<Serde, Serde> kvSerdes = getKVSerdes(streamId, serde);
+
+    boolean isKeyed;
+    KV<Serde, Serde> kvSerdes;
+    if (streamSerde == null) { // if no explicit or default system descriptor serde available
+      isKeyed = true; // assume keyed stream
+      kvSerdes = new KV<>(null, null); // and that key and msg serdes are provided for job.default.system in configs
```

Comment:
Should this be KV(NoOpSerde, NoOpSerde)?

Reply:
No, this should be (null, null). There's a difference in how NoOpSerde and null serde value is handled. When generating configs, a NoOpSerde value is serialized as the serde value, it's then ignored in the container. A null serde value is not added to the generated configs at all. Due to how config resolution works (generated config overrides user config), setting NoOpSerde here will actually override the key/msg serde for job.default.system.

Reply:
If "job.default.system" is not specified, then will it fall back to KV(NoOpSerde, NoOpSerde) after that?

Reply:
Not anywhere in code, but it'll fall back to the same behavior - no serde for incoming/outgoing messages. Most likely this will result in ClassCastExceptions at runtime.